### PR TITLE
LPS-80214 Introduce ConfigurationScreen as a way to provide a custom configuration UI within System Settings

### PR DESCRIPTION
**Pull request by JorgeFerrer**
Hey @drewbrokke, this is the second part of https://issues.liferay.com/browse/LPS-75333. I've tried to simplify it the most I could so that it's dead simple to review. Let me know if you have any doubt. BTW, I've included a last commit with the sample screen that I have created for testing, so that you can use it to, but make sure to not send it to Bchan /cc @pei-jung

### CODE REVIEW

**Comment by JorgeFerrer** [comment]
ci:test

**Comment by JorgeFerrer** [comment]
ci:retest

**Comment by JorgeFerrer** [comment]
ci:retest

**Comment by pei-jung** [comment]
Hey @JorgeFerrer,

Thanks for working on this! We took a quick look at the changes, and it seems that it only allows adding new entries to the configuration UI. What happens if an user tries to override an existing configuration with the ConfigurationScreen? It looks like it will end up with two entries in the menu, one with the default UI from ConfigurationMdoels and another with the custom UI from ConfigurationScreens (https://github.com/JorgeFerrer/liferay-portal/blob/27dddfc11d582040a09360889f9449189846fb27/modules/apps/foundation/configuration-admin/configuration-admin-web/src/main/java/com/liferay/configuration/admin/web/internal/util/ConfigurationModelRetrieverImpl.java#L196-L227).

Is that intentional? or are we missing something here?

**Comment by JorgeFerrer** [comment]
Hey @pei-jung ,

That's a good question. The main use case I was trying to support was creating completely new UIs, which are not necessarily backed by a ConfigurationModel. This will provide an alternative to having to develop custom portlets.

And for those cases where the custom UI still uses configuration models, the creation of an automatic UI can be avoided by not using Liferay's ExtendedObjectClassDefinition. 

This is something that is important to explain in the documentation. I'll add a comment to JIRA later today with a longer explanation. I will also think through if it might be better to make it more explicit, for example by adding an attribute to ExtendedObjectClassDefinition (generateUI?)

**Comment by drewbrokke** [comment]
@JorgeFerrer So you're saying that adding the `ExtendedObjectClassDefinition` is an implicit opt-in for a generated interface?  Does it have any other uses besides display purposes?  If so, then I think a `generateUI` attribute would be useful.  If not, then I think it wouldn't be necessary and we could just make sure to document its purpose clearly.

**Comment by drewbrokke** [comment]
The other thought I have is that ConfigurationScreen entries should somehow be indexed.  As it is right now, they are not indexed and so they're not searchable, which from my experience is almost essential to navigate this portlet quickly.  Even just indexing the name alone would provide a huge benefit.

File: modules/apps/foundation/configuration-admin/configuration-admin-web/src/main/java/com/liferay/configuration/admin/web/internal/portlet/action/ViewConfigurationScreenMVCRenderCommand.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	private BundleContext _bundleContext;
+
+	@Reference
+	private ConfigurationModelRetriever _configurationModelRetriever;
+
+	@Reference
+	private Portal _portal;
+
+	private final Map<String, MVCRenderCommand> _renderCommands =
+		new HashMap<>();
```
[the remark this answers is not in the document]
Comment (JorgeFerrer) [reply modules/apps/foundation/configuration-admin/configuration-admin-web/src/main/java/com/liferay/configuration/admin/web/internal/portlet/action/ViewConfigurationScreenMVCRenderCommand.java]: Yes, good catch!

**Comment by JorgeFerrer** [comment]
@drewbrokke good point again. We should probably index any types of ConfigurationEntry. I'll look into it.

**Comment by drewbrokke** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Pull request closed without merging**